File: 04-1.py
#!/usr/bin/env python

import re

lines = [row.strip() for row in open('04.input').readlines()]

card_format = re.compile("^Card +(?P<num>[0-9]+): (?P<winning>[0-9 ]*)\\|(?P<have>[0-9 ]*)$")

score = 0
for line in lines:
    m = card_format.match(line)
    winning = set([int(n) for n in m['winning'].split()])
    have = set([int(n) for n in m['have'].split()])

    if len(winning & have) > 0:
        score += 2 ** (len(winning & have) - 1)
    else:
        score += 0

print(score)

File: 04-1.rb
#!/usr/bin/env ruby

lines = File.read('04.input').lines.map(&:strip)

card_format = /^Card +(?<num>[[:digit:]]+): (?<winning>[[:digit:] ]*)\|(?<have>[[:digit:] ]*)$/

score = lines.sum do |line|
  m = line.match card_format
  winning = m['winning'].split.map(&:to_i)
  have = m['have'].split.map(&:to_i)
  if (winning & have).size > 0
    2 ** ((winning & have).size - 1)
  else
    0
  end
end

print score, "\n"
